File: src/tools/guard.ts
import { sanitize } from '../agents/prompt/builder.js';
import { trimToBudget } from '../agents/prompt/budget.js';
import { logger } from '../infra/logger.js';

/**
 * Lop bao ve cho ket qua cong cu.
 *
 * ARCHITECTURE.md section 7.2 chi tinh injection qua tai lieu do ADMIN nap. Web
 * search dua vao van ban do NGUOI LA soan, va trong vong ReAct thi observation do
 * lai quyet dinh hanh dong tiep theo — injection duoc khuech dai.
 *
 * CO Y KHONG CHAN theo tu khoa. Chan bang danh sach tu vua de vuot (viet lai mot
 * chut la lot), vua tao cam giac an toan gia khien nguoi ta bo qua cac lop that su
 * co tac dung: boc the, noi ro trong RULES rang do la du lieu, va khong bao gio
 * dua ket qua cong cu vao role 'system'.
 *
 * Ham nay ton tai de AUDIT: biet co ai dang thu, va thu bang cach nao.
 */
const INJECTION_PATTERNS: readonly (readonly [name: string, pattern: RegExp])[] = [
  ['bo-qua-huong-dan', /\b(bỏ qua|phớt lờ|quên)\s+(mọi\s+)?(hướng dẫn|chỉ dẫn|quy tắc|luật)/i],
  ['ignore-instructions', /\bignore\s+(all\s+|previous\s+|prior\s+|above\s+)+(instructions?|rules?)/i],
  ['doi-vai', /\b(bây giờ|từ giờ)\s+(bạn|mày)\s+(là|thành)\b/i],
  ['you-are-now', /\byou\s+are\s+now\b|\bact\s+as\s+(a\s+)?(dan|jailbreak)/i],
  ['lo-system-prompt', /\b(in ra|hiển thị|tiết lộ|cho xem)\s+.{0,20}(system prompt|prompt hệ thống)/i],
  ['reveal-prompt', /\b(reveal|print|show|repeat)\s+.{0,20}(system prompt|your instructions)/i],
  ['che-do-dac-biet', /\b(chế độ|mode)\s+(nhà phát triển|developer|god|unrestricted)/i],
  ['gia-danh-quan-tri', /\b(tôi là|mình là)\s+(quản trị viên|admin|người tạo ra bạn)/i],
];

export type InjectionScan = Readonly<{ suspicious: boolean; patterns: readonly string[] }>;

export function detectInjection(text: string): InjectionScan {
  const hits = INJECTION_PATTERNS.filter(([, pattern]) => pattern.test(text)).map(([name]) => name);
  return { suspicious: hits.length > 0, patterns: hits };
}

/**
 * Boc ket qua cong cu thanh mot khoi du lieu an toan de dua vao prompt.
 *
 * Ba viec, theo dung thu tu:
 *   1. sanitize()  — bo the dong gia, chan noi dung tu thoat khoi hop.
 *   2. cat theo tran tang 'tool' — mot trang web dai khong duoc nuot ca cua so.
 *   3. boc trong <ket_qua_cong_cu> — RULES trong system prompt noi ro day la
 *      DU LIEU, khong phai chi thi.
 */
export function wrapObservation(args: {
  toolName: string;
  source: string;
  content: string;
  traceId: string;
}): string {
  const scan = detectInjection(args.content);
  if (scan.suspicious) {
    logger.warn(
      { traceId: args.traceId, tool: args.toolName, source: args.source, patterns: scan.patterns },
      'ket qua cong cu chua mau giong prompt injection — ghi nhan de audit, KHONG chan',
    );
  }

  const clean = trimToBudget(sanitize(args.content), 'tool');
  if (clean.trimmedTokens > 0) {
    logger.info(
      { traceId: args.traceId, tool: args.toolName, trimmedTokens: clean.trimmedTokens },
      'ket qua cong cu vuot tran, da cat',
    );
  }

  const flag = scan.suspicious ? ' canh_bao="chua_cau_ra_lenh"' : '';
  return `<ket_qua_cong_cu cong_cu="${args.toolName}" nguon="${args.source}"${flag}>\n${clean.text}\n</ket_qua_cong_cu>`;
}

File: src/tools/guard.py
"""Lop bao ve cho KET QUA CONG CU — mot trong ba be mat dung chung bo do injection.

Bo do da chuyen sang agents/policy/injection.py (08/09/2026) de ca ba be mat dung
chung mot danh sach mau: tin nhan nguoi dung, tai lieu luc nap, va ket qua cong cu.
Ba danh sach roi nhau la ba danh sach se lech nhau sau vai lan sua.

Phan con lai o day la thu rieng cua ket qua cong cu: boc the, cat tran.

ARCHITECTURE.md section 7.2 chi tinh injection qua tai lieu do ADMIN nap. Web search
dua vao van ban do NGUOI LA soan, va trong vong ReAct thi observation do lai quyet
dinh hanh dong tiep theo — injection duoc khuech dai.

CO Y KHONG CHAN theo tu khoa. Chan bang danh sach tu vua de vuot (viet lai mot chut
la lot), vua tao cam giac an toan gia khien nguoi ta bo qua cac lop that su co tac
dung: boc the, noi ro trong RULES rang do la du lieu, va khong bao gio dua ket qua
cong cu vao role 'system'.

Ham detect_injection ton tai de AUDIT: biet co ai dang thu, va thu bang cach nao.
"""

from agents.policy.injection import detect_injection
from agents.ports.logger import LoggerPort
from agents.prompt.budget import exceeds_budget, trim_to_budget
from agents.prompt.builder import sanitize


def exceeds_tool_budget(text: str) -> bool:
    """Ket qua co vuot tran tang 'tool' khong.

    Cho goi dung cai nay de biet co CAN NEN khong. Nen bang model re giu duoc trich
    dan; cat cung thi mat phan cuoi — ma phan cuoi thuong la noi chua DOI va link
    nguon, dung thu bat buoc phai trich dan.
    """
    return exceeds_budget(text, "tool")


def wrap_observation(
    *,
    tool_name: str,
    source: str,
    content: str,
    trace_id: str,
    logger: LoggerPort | None = None,
) -> str:
    """Boc ket qua cong cu thanh mot khoi du lieu an toan de dua vao prompt.

    Ba viec, theo dung thu tu:
      1. sanitize()  — bo the dong gia, chan noi dung tu thoat khoi hop.
      2. cat theo tran tang 'tool' — mot trang web dai khong duoc nuot ca cua so.
      3. boc trong <ket_qua_cong_cu> — RULES trong system prompt noi ro day la
         DU LIEU, khong phai chi thi.

    logger tiem vao chu khong import infra/logger: import se keo theo config/, va luc
    do unit test cua chinh ham nay lai doi du bien moi truong — dung cai ma L2 sinh
    ra de tranh.
    """
    scan = detect_injection(content)
    if scan.suspicious and logger is not None:
        logger.warning(
            "ket qua cong cu chua mau giong prompt injection — ghi nhan de audit, KHONG chan",
            trace_id=trace_id,
            tool=tool_name,
            source=source,
            patterns=list(scan.patterns),
        )

    # Cat cung o day la LUOI CUOI CUNG — cho goi da thu nen bang model re truoc do.
    clean = trim_to_budget(sanitize(content), "tool")
    if clean.trimmed_tokens > 0 and logger is not None:
        logger.warning(
            "ket qua cong cu VAN vuot tran sau khi nen — cat cung, co the mat trich dan o cuoi",
            trace_id=trace_id,
            tool=tool_name,
            trimmed_tokens=clean.trimmed_tokens,
        )

    flag = ' canh_bao="chua_cau_ra_lenh"' if scan.suspicious else ""
    return (
        f'<ket_qua_cong_cu cong_cu="{tool_name}" nguon="{source}"{flag}>\n'
        f"{clean.text}\n</ket_qua_cong_cu>"
    )
